### PR TITLE
MR-3446: Migrate fetchHealthPlanPackage resolver to use new db functions

### PR DESCRIPTION
**Pull request by JasonLin0991**
## Summary
[MR-3446](https://qmacbis.atlassian.net/browse/MR-3446)

This work adds logic behind the `rates-db-refactor` flag that switches the `fetchHealthPlanPackage` resolver to get Contract data from new tables using db functions `findContractWithHistory` and `findDraftContract`. We can do create a new submission using the new Postgres tables, still cannot update though.

- We are using both `findContractWithHistory` and `findDraftContract` to build a `DRAFT` contract revision data and its revision history. `SUBMITTED` contract will skip `findDraftContract`.
- This resolver has been changed from returning an `undefined` HPP to throwing an error if Postgres cannot find the package for the provided ID.
   - A not found contract will also throw an error on `undefined` results. 
      - `findContractWithHistory` is changed to throw an error if a contract is not found. It's converted to a `StoreError` before returning. `submitContract`, `unlockContract`, and `updateDraftContract` all have been updated to do the same for records not found.
      - Updated the front end to handle the thrown error instead of an undefined package and updated tests.
      - `StateSubmissionForm`, `SubmissionSideNav`, and `SubmissionRevisionSummary` components were updated.
- `storeError.ts`, I've changed the return code and message for Prisma error code `2025`. In the error, this code is a `NotFoundError` and feels more appropriate now that `findContractWithHistory` can also throw this error. 
- Tests for `fetchHealthPlanPackage` resolver are not fully covered for the `rates-db-refactor` flag on because we still need to migrate update, submit, and unlock resolvers. Once they are migrated we should be able to fully cover tests with flag on like we did with `createHealthPlanPackage` resolver.
- Fixed a bug in `testLDService` where the `defaultFeatureFlags` were mutated with the flags passed in and becoming persistent between tests. The fix was to copy `defaultFeatureFlags` to `featureFlags` instead of referencing it.
- If we are ok with adding tests for `fetchHealthPlanPackage` as new resolvers are migrated then I will add them as a AC for these three tickets [MR-3447](https://qmacbis.atlassian.net/browse/MR-3447), [MR-3448](https://qmacbis.atlassian.net/browse/MR-3448), and [MR-3436](https://qmacbis.atlassian.net/browse/MR-3436) before merging this work in.
   - [x] Add and AC to update `fetchHealthPlanPackage` tests to [MR-3447](https://qmacbis.atlassian.net/browse/MR-3447), [MR-3448](https://qmacbis.atlassian.net/browse/MR-3448), and [MR-3436](https://qmacbis.atlassian.net/browse/MR-3436)

#### Test cases covered

- `fetchHealthPlanPackage.test.ts`
   - `'returns nothing if the ID doesnt exist'` -> `'returns error if the ID doesnt exist'`
   - `'fetchHealthPlanPackage rates-db-refactor flag on tests'`
      - `'returns package with one revision'`
         - Only testing a draft contract we do not have the function to convert draft revision to submitted revision yet. I think this will come when we migrate `submitHealthPlanPackage` resolver.
      - `'returns error if the ID doesnt exist'`

## QA guidance

- Test various submission options on the new submission form to ensure there are no strange edge cases.
- I've changed how the front end handles packages that are not found. Previously the server would return `pkg` as `undefined`, and our frontend code would render the `Error404` component. Now that we are throwing an error from the server with the code `NOT_FOUND`, the front-end code checks the responses error code for `NOT_FOUND`, then renders the `Error404` component.
   - Please test that using invalid submission ids in the URL still renders the 404 pages.
   - Test for regressions?

### CODE REVIEW

**Review by macrael — CHANGES_REQUESTED** [review]
Yay I'm excited for this all to come together. I def would like to drop the StoreError stuff though, I wrote in a comment how to use custom js errors instead but am happy to chat about what that would look like in detail as well.

File: services/app-api/src/postgres/contractAndRates/findContractWithHistory.ts [hunk 1]
```diff
@@ -10,9 +11,9 @@ import { updateInfoIncludeUpdater } from '../prismaHelpers'
 async function findContractWithHistory(
     client: PrismaTransactionType,
     contractID: string
-): Promise<ContractType | Error> {
+): Promise<ContractType | StoreError | Error> {
```
[the remark this answers is not in the document]
Comment (macrael) [reply services/app-api/src/postgres/contractAndRates/findContractWithHistory.ts:14]: This strategy is documented in an ADR: https://github.com/Enterprise-CMCS/managed-care-review/blob/main/docs/architectural-decision-records/012-custom-error-types.md

Note that in that ADR it shows calling Object.setPrototypeOf(this, FooError.prototype); instead of setting name, I'd do some googling to see what is required to do this right.

File: services/app-api/src/resolvers/healthPlanPackage/fetchHealthPlanPackage.ts [hunk 2]
```diff
@@ -5,43 +5,151 @@ import {
...
-            logError('fetchHealthPlanPackage', errMessage)
-            setErrorAttributesOnActiveSpan(errMessage, span)
-            return {
-                pkg: undefined,
+        const ratesDatabaseRefactor = await launchDarkly.getFeatureFlag(
+            context,
+            'rates-db-refactor'
+        )
+
+        let pkg: HealthPlanPackageType = <HealthPlanPackageType>{}
```
[the remark this answers is not in the document]
Comment (JasonLin0991) [reply services/app-api/src/resolvers/healthPlanPackage/fetchHealthPlanPackage.ts:35]: It's a way of type assertion. This was left over from a previous version of the work; we don't need this anymore.

File: services/app-api/src/resolvers/healthPlanPackage/fetchHealthPlanPackage.ts [hunk 3]
```diff
@@ -5,43 +5,151 @@ import {
...
+                    extensions: {
+                        code: 'INTERNAL_SERVER_ERROR',
+                        cause: 'PROTO_DECODE_ERROR',
+                    },
+                })
+            }
+
+            pkg = convertedPkg
+        } else {
+            const result = (await store.findHealthPlanPackage(input.pkgID)) as
```
[the remark this answers is not in the document]
Comment (JasonLin0991) [reply services/app-api/src/resolvers/healthPlanPackage/fetchHealthPlanPackage.ts:129]: Also left over from previous work. Not needed anymore.

File: services/app-api/src/postgres/contractAndRates/findContractWithHistory.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (JasonLin0991) [reply services/app-api/src/postgres/contractAndRates/findContractWithHistory.ts:14]: Thanks! I'll make these changes.

File: services/app-api/src/testHelpers/launchDarklyHelpers.ts [hunk 4]
```diff
@@ -7,7 +7,7 @@ import {
 import { defaultFeatureFlags } from '../launchDarkly/launchDarkly'
 
 function testLDService(mockFeatureFlags?: FeatureFlagSettings): LDService {
-    const featureFlags = defaultFeatureFlags
+    const featureFlags = Object.assign({}, defaultFeatureFlags)
```
[the remark this answers is not in the document]
Comment (JasonLin0991) [reply services/app-api/src/testHelpers/launchDarklyHelpers.ts:10]: We are using that `defaultFeatureFlags` object to derive types. I don't know how much of a refactor it would take if we change it to a function and try to get types from it. So I opted to copy it instead. I would like to split it into a new ticket if turning it into a function is the way we should go though.

[the remark this answers is not in the document]
Comment (macrael) [reply services/app-api/src/testHelpers/launchDarklyHelpers.ts:10]: def don't need to change this, just seeing you solve the same problem we saw elsewhere

File: services/app-api/src/errors/errors.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+class NotFoundError extends Error {
```
Comment (macrael) [inline services/app-api/src/errors/errors.ts:1]: I think this error can live inside the postgres package and be exported from there, I see it as part of the contract of that package, since some of those functions can return it. 

You may ask, what makes this different from a domain model then, and I don't have a great answer, except that I don't see us reusing the not found error for other packages? I could be wrong.

File: services/app-api/src/postgres/contractAndRates/findContractWithHistory.ts [hunk 6]
```diff
@@ -11,9 +11,9 @@ import { convertPrismaErrorToStoreError, StoreError } from '../storeError'
 async function findContractWithHistory(
     client: PrismaTransactionType,
     contractID: string
-): Promise<ContractType | StoreError | Error> {
```
Comment (macrael) [inline services/app-api/src/postgres/contractAndRates/findContractWithHistory.ts]: Open question wether this should be `| NotFoundError` I'm fine with it not for now but it should probably be documented somewhere, maybe in the comment above?

File: services/app-api/src/postgres/contractAndRates/findContractWithHistory.ts [hunk 7]
```diff
@@ -46,10 +46,16 @@ async function findContractWithHistory(
...
+        if (!contract) {
+            const err = `PRISMA ERROR: Cannot find contract with id: ${contractID}`
+            console.error(err)
+            return new NotFoundError(err)
+        }
+
         return parseContractWithHistory(contract)
     } catch (err) {
         console.error('PRISMA ERROR', err)
-        return convertPrismaErrorToStoreError(err)
```
Comment (macrael) [inline services/app-api/src/postgres/contractAndRates/findContractWithHistory.ts]: Are there any other store errors we should be returning instead of just the raw prisma error? I'm not sure we handle any of them so we can leave it probably.

**Review by macrael — APPROVED** [review]
Looks great

File: services/app-api/src/postgres/contractAndRates/findContractWithHistory.ts (hunk 6 above)
Comment (JasonLin0991) [reply services/app-api/src/postgres/contractAndRates/findContractWithHistory.ts]: I think it should, I just forgot to add it to the return type. My reason for adding it is because we are specifially handling not found errors and returning `NotFoundError`.

File: services/app-api/src/postgres/contractAndRates/findContractWithHistory.ts (hunk 7 above)
Comment (JasonLin0991) [reply services/app-api/src/postgres/contractAndRates/findContractWithHistory.ts]: I could not think of any others. I'm also good with leaving it and if others creep up, we can add them. I think the main factor for making a custom error is if we are checking for specific errors in the resolvers, like `NotFoundError`.

File: services/app-api/src/errors/errors.ts (hunk 5 above)
Comment (JasonLin0991) [reply services/app-api/src/errors/errors.ts:1]: That sounds good to me!
